tests/test_views.py

```diff
@@ -3605,10 +3614,193 @@ def test_confirmation_link_registers_user(self):
+            )
+
+            # campaign register: user with auth
+            data = login_and_register_handler(self.auth, login=False, campaign=campaign, next_url=None, logout=False)
+            assert_equal(data.get('status_code'), http.FOUND)
+            assert_equal(data.get('next_url'), campaign_url_for(campaign))
+
+            # campaign register: user without auth
+            data = login_and_register_handler(self.no_auth, login=False, campaign=campaign, next_url=None, logout=False)
+            assert_equal(data.get('status_code'), http.OK)
+            if is_native_login(campaign):
+                # native campaign: prereg and erpc
+                assert_equal(data.get('next_url'), campaign_url_for(campaign))
+            elif is_proxy_login(campaign):
+                # proxy campaign: preprints and branded ones
+                assert_equal(
+                    data.get('next_url'),
+                    web_url_for('auth_login', next=campaign_url_for(campaign), _absolute=True)
+                )
+
+    def test_campaign_login_and_register_with_next_url(self):
+
+        for campaign in CAMPAIGNS:
+
+            if is_institution_login(campaign):
+                continue
+
+            next_url = campaign_url_for(campaign)
+
+            # campaign login: user with auth
+            data = login_and_register_handler(self.auth, login=True, campaign=campaign, next_url=next_url, logout=False)
+            assert_equal(data.get('status_code'), http.FOUND)
+            assert_equal(data.get('next_url'), next_url)
+
+            # campaign login: user without auth
+            data = login_and_register_handler(self.no_auth, login=True, campaign=campaign, next_url=next_url, logout=False)
+            assert_equal(data.get('status_code'), http.FOUND)
+            assert_equal(
+                data.get('next_url'),
+                web_url_for('auth_register', campaign=campaign, next=next_url)
+            )
+
+            # campaign register: user with auth
+            data = login_and_register_handler(self.auth, login=False, campaign=campaign, next_url=next_url, logout=False)
+            assert_equal(data.get('status_code'), http.FOUND)
+            assert_equal(data.get('next_url'), next_url)
+
+            # campaign register: user without auth
+            data = login_and_register_handler(self.no_auth, login=False, campaign=campaign, next_url=next_url, logout=False)
+            assert_equal(data.get('status_code'), http.OK)
+            if is_native_login(campaign):
+                # native campaign: prereg and erpc
+                assert_equal(data.get('next_url'), next_url)
+            elif is_proxy_login(campaign):
+                # proxy campaign: preprints and branded ones
+                assert_equal(
+                    data.get('next_url'),
+                    web_url_for('auth_login', next= next_url, _absolute=True)
+                )
+
+    def test_login_and_register_with_logout_flag(self):
```

Comment:
A special test for `claim_user_register()` when user decides to claim under another account or when the user is already a contributor.